text/0000-add-logical-operators.md

```diff
@@ -0,0 +1,95 @@
+or the right-most arguments if **all** evaluate to _truthy_.
+This is *NOT* equivalent to the `{{and}}` helper from `ember-truth-helpers` because unlike this proposed helper, the one in `ember-truth-helpers`
+uses Javascript's definition of truthiness.
+
+#### `{{or}}`
+Takes at least two positional arguments, bound or unbound. Raises an error if invoked with less than two arguments.
+It evaluates arguments left to right, returning the first one that is _truthy_ (**by handlebar's definition of truthiness**) or the
+right-most argument if all evaluate to _falsy_.
+This is *NOT* equivalent to the `{{or}}` helper from `ember-truth-helpers` because unlike this proposed helper, the one in `ember-truth-helpers`
+uses Javascript's definition of truthiness.
+
+#### `{{not}}`
+Unary operator. Raises an error if invoked with more than one positional argument. If the given value evaluates to a _truthy_ value (**by handlebar's definition of truthiness**),
+the `false` is returned. If the given value evaluates to a _falsy_ value (**by handlebar's definition of truthiness**) then it returns `true`.
+This is *NOT* equivalent to the `{{not}}` helper from `ember-truth-helpers` because unlike this proposed helper, the one in `ember-truth-helpers`
+uses Javascript's definition of truthiness.
+
+#### Handlebar's definition of truthiness
+This is the most important detail of this proposal because it's where it deviates from `ember-truth-helpers`.
+Handlebars has its own definition of _truthyness_, which is similar to Javascripts except that empty arrays are
+considered **falsy**, while in JS are considered **truthy**.
+
+
+This RFC intentionally leaves the implementation details unspecified, but one can think of those helpers as macros that
+expand to combinations of `if`s.
+
+##### `{{and}}`
+- `{{and a b}}` is equivalent to `{{if a b a}}`
+- `{{and a b c}}` is equivalent to  `{{if a (if b c b) a}}`
+- and so on
+
+##### `{{or}}`
+- `{{or a b}}` is equivalent to `{{if a a b}}`
+- `{{and a b c}}` is equivalent to  `{{if a a (if b b c)}}`
+- and so on
+
+##### `{{not}}`
+- `{{not a}}` is equivalent to `{{if a false true}}`
+
+## How we teach this
+
+The introduction of these helpers does not impact the current mental model for Ember applications.
+
+In addition to API and Guides documentation with illustrative examples of usage of the various helpers, stressing out
+where the definition of truthiness of handlebars is different from the one in Javascript.
+
+## Drawbacks
+
+Adding new helpers increases the surface area of the framework and the code the core team commits to support long term.
+
+## Alternatives
+
+One alternative path is to not take any action and let users continue to define their own helpers (or install `ember-truth-helpers`).
+
+## Unresolved questions
+
+- Consider following Javascript's definition of truthiness. This would also help with the transition from `ember-truth-helpers`.
```

Comment:
Probably belongs under “Alternatives” instead.

Reply:
I'd argue that it belongs here. The way I understand RFCs, alternatives is where you list alternative options to implementing the RFC, not so much about doubts regarding what the best implementation is.

If anyone can confirm that I got it wrong, I'm happy to change it.

Reply:
I think an exact implementation of ember-truth-helpers logical operators (except maybe not) is not what anyone wants, given the desire for short circuit logic. But alternatives sections are not always fully specced out, so I think with just a little more detail, you could write an alternative.

Reply:
Being someone who believes that the lack of logical operators is the main benefit that handlebars provides, and one who has never had a need to install this addon in the many ember apps i have built, simply assuming that this is something that is required to be installed does not sit well with me. I would like somewhere in this process to highlight the benefits that not having these be a part of the framework provides.

Reply:
@webark You can certainly live without those helpers, but I'll add some examples of things that are much nicer with helpers like those:

## Default Values in template-only components
Imagine a template only component that wants to have a default value for say, `@name`

```hbs
<img class="avatar" src={{or @user.avatar "/imgs/default-avatar.png">
```

## Perform an operation on each iteration of a loop without extracting another component
```hbs
{{#each @users as |user|}}
  Name: {{user.name}} 
  {{#if (and session.hasAdminPrivileges (not user.inactive))}}
    <button {{on "click" this.activate}}>Delete</button>
  {{/if}}
{{/each}}
```

In both situation you can find a way to do the same things without using any of those helpers, but it involve either create a JS file for a component only to put a computed property on it, or to extract a new component to call it on each iteration of a loop, just to perform a very simple logical operation.

It's a matter of convenience and as any tool it can be abused, but the popularity of ember-truth-helpers proves that devs really find it convenient.

I think that Ember providing just the most basic and used ones (`and`, `or`, `not`, `eq` are the main 4 for me, but this RFC focuses on the first 3) and optimized at the VM level we are alleviating a pain.

Reply:
call me a curmudgeon (and i’ll postpone my “at this point we might as well be using jsx” rant for another day) I just feel it should be mentioned why this was never in core to begin with. Your examples, to me, are the base examples of why. Just a brief “handlebars was originally meant to be “logic-less” but it has grown, and the community has pushed it, past that for sometime now” blurb would be helpful, at least for a historical sense.